packages/sanity/src/core/i18n/types.ts

```diff
@@ -0,0 +1,161 @@
+   */
+  locales?: LocalesOption
+
+  /**
+   * Bundles contain "resources" (strings) that yields translations for different locales
+   * throughout the studio. The strings are scoped to a specific locale and namespace.
+   * Namespaces in this context usually means a specific part of the studio, like a tool or plugin.
+   */
+  bundles?: LocalesBundlesOption
+}
+
+/**
+ * A locale resource bundle where the locale is inherited from the parent locale definition.
+ *
+ * @beta
+ */
+export type ImplicitLocaleResourceBundle = Omit<LocaleResourceBundle, 'locale'>
+
+/**
+ * A collection of locale resources for a given locale and namespace.
+ * In other words, an object of translated locale strings.
+ *
+ * @beta
+ */
+export interface LocaleResourceBundle {
+  /**
+   * The locale ID the resources belong to, eg `en-US`, `nb-NO`, `th-TH`…
+   */
+  locale: string
+
+  /**
+   * The namespace the resources belong to, eg `vision`, `desk`, `studio`…
+   */
+  namespace: string
+
+  /**
+   * An object of locale resources, or a function that resolves to one.
+   * The localization framework automatically handles ESM modules with a default export,
+   * since a common use case is to dynamically load a resource file on use. This is the
+   * preferred option, since it allows for lazy loading of locale resources on use.
+   */
+  resources:
+    | LocaleResourceRecord
+    | (() => Promise<LocaleResourceRecord | {default: LocaleResourceRecord}>)
+
+  /**
+   * Whether the resources should be merged deeply (eg for nested objects). Default: true
+   */
+  deep?: boolean
+
+  /**
+   * Whether any existing resource keys for the namespace be overwritten. Default: true
+   */
+  overwrite?: boolean
+}
+
+/**
+ * A locale definition, which describes a locale and its resources.
+ *
+ * @beta
+ */
+export interface LocaleDefinition {
+  /**
+   * The ID of the locale, eg `en-US`, `nb-NO`, `th-TH`…
+   */
+  id: string
+
+  /**
+   * The title of locale, eg `English (US)`, `Norsk (bokmål)`, `ไทย`…
+   */
+  title: string
+
+  /**
+   * Array of resource bundles for this locale, if any.
+   *
+   * Generally you'll want to provide some base resources, eg for the studio core namespace,
+   * as well as for common namespaces like `desk` and `vision`. You can also provide resources
+   * for other plugins/namespaces - but preferably the resources should be provided as an async
+   * function that imports the resources, in order to lazy load them on use.
+   */
+  bundles?: (ImplicitLocaleResourceBundle | LocaleResourceBundle)[]
+
+  // @todo allow fallback locales? eg [no-nn, no-nb, en]
+}
+
+/** @public */
+export interface LocaleSource {
+  /**
+   * Current locale ID (eg `en-US`, `nb-NO`, `th-TH`…)
+   */
+  currentLocale: string
+
+  /**
+   * Array of locale definitions
+   */
+  locales: {id: string; title: string}[]
+
+  /**
+   * Translation function, eg `t('some.key') => 'Some string'`
+   */
+  t: TFunction
```

Comment:
the TFunction signature looks pretty complex - do you think it could make it difficult to move to another i18n backend in the future? If so, would it make sense to expose a simpler but compatible signature here?

Reply:
Ah - I see now that you already mentioned this. Feel free to disregard while I learn how to read PR descriptions.